finality-provider/service/rollup_rand_committer.go

```diff
@@ -0,0 +1,238 @@
+	activationBlkHeight, err := rrc.consumerCon.QueryFinalityActivationBlockHeight(ctx)
+	if err != nil {
+		return false, 0, fmt.Errorf("failed to query finality activation block height: %w", err)
+	}
+
+	// Calculate tip height with delay
+	tipHeightWithDelay := tipBlock.GetHeight() + uint64(rrc.cfg.TimestampingDelayBlocks) // #nosec G115
+
+	// ROLLUP-SPECIFIC: Determine startHeight with interval awareness from the beginning
+	var alignedStartHeight uint64
+	switch {
+	case lastCommittedHeight < tipHeightWithDelay:
+		// Need to start from tipHeightWithDelay, but align it to voting schedule
+		baseHeight := max(tipHeightWithDelay, activationBlkHeight)
+		alignedStartHeight = rrc.calculateFirstEligibleHeightWithActivation(baseHeight, activationBlkHeight)
+
+	case rrc.needsMoreVotingRandomness(lastCommittedHeight, tipHeightWithDelay, activationBlkHeight):
+		// Need to continue from where we left off, but with interval spacing
+		// For sparse generation, we need to check if we have enough *voting* heights covered
+		baseHeight := max(lastCommittedHeight+rrc.interval, activationBlkHeight)
+		alignedStartHeight = rrc.calculateFirstEligibleHeightWithActivation(baseHeight, activationBlkHeight)
+
+	default:
+		// Check if we have sufficient *voting* randomness, not just any randomness
+		// Calculate the last voting height we have randomness for
+		lastVotingHeight := rrc.getLastVotingHeightWithRandomness(lastCommittedHeight, activationBlkHeight)
+		requiredVotingHeight := tipHeightWithDelay + uint64(rrc.cfg.NumPubRand)*rrc.interval
+
+		if lastVotingHeight >= requiredVotingHeight {
+			// Sufficient voting randomness, no need to commit
+			rrc.logger.Debug(
+				"the rollup finality-provider has sufficient voting randomness, skip committing more",
+				zap.String("pk", rrc.btcPk.MarshalHex()),
+				zap.Uint64("tip_height", tipBlock.GetHeight()),
+				zap.Uint64("last_committed_height", lastCommittedHeight),
+				zap.Uint64("last_voting_height", lastVotingHeight),
+				zap.Uint64("required_voting_height", requiredVotingHeight),
+			)
+
+			return false, 0, nil
+		}
+
+		// Need more voting randomness
+		baseHeight := max(lastCommittedHeight+rrc.interval, activationBlkHeight)
+		alignedStartHeight = rrc.calculateFirstEligibleHeightWithActivation(baseHeight, activationBlkHeight)
+	}
+
+	rrc.logger.Debug(
+		"the rollup finality-provider should commit randomness",
+		zap.String("pk", rrc.btcPk.MarshalHex()),
+		zap.Uint64("tip_height", tipBlock.GetHeight()),
+		zap.Uint64("last_committed_height", lastCommittedHeight),
+		zap.Uint64("aligned_start_height", alignedStartHeight),
+		zap.Uint64("interval", rrc.interval),
+	)
+
+	return true, alignedStartHeight, nil
+}
+
+// needsMoreVotingRandomness determines if we need more voting randomness
+// by checking if our last committed voting height can cover the required buffer
+func (rrc *RollupRandomnessCommitter) needsMoreVotingRandomness(lastCommittedHeight, tipHeightWithDelay, activationHeight uint64) bool {
```

Comment:
All of those caluculations here should have corresponding unit tests. For now I will trust all of those return correct results.